### ISSUE TITLE
CUDA error: an illegal memory access was encountered

### ISSUE THREAD

**Issue by algorithmx** [JuliaGPU/CUDAnative.jl#160]
Hi everyone, 

I encountered an error which I had no clue on how to git over it. I hope that someone can help me.

The code is the following. I am trying to write a cuda native function `kernel_zip_V1`  in Julia to "zip" two vectors in a particular manner. At the bottom lines of the code I compared the serial computation and the GPU computation and always find agreements of the results.

------------------------------------------------------------------------------
------------------------------------------------------------------------------
### CUDA kernel
`function kernel_zip_V1(psi, psiconj, val_i, perm_i, res)`
`    i = (blockIdx().x-1) * blockDim().x + threadIdx().x`
`    res[i] = psiconj[i]* psi[perm_i[i]] * val_i[i]`
`    return nothing`
`end`
### set up device
`dev = CuDevice(0)`
`name(dev)`
`ctx = CuContext(dev)`
### generate some data
`len = 1024`
`perm = Int32.(collect(1:100*len))`
`val = rand(Complex128, 100*len)`
`psi = rand(Complex128, 100*len)`
`psi = psi/norm(psi)`
`psic = conj.(psi)`
### allocate & upload on the GPU
`perm_dev = CuArray(perm)`
`val_dev = CuArray(val)`
`psi_dev = CuArray(psi)`
`psiconj_dev = CuArray(psic)`
`res = similar(psi_dev)`
### execute and fetch results
# there is a pRoblem here!
### always have "CUDA error: an illegal memory access was encountered" message with failure
`@cuda (100,1024) kernel_zip_V1(psi_dev, psiconj_dev, val_dev, perm_dev, res)`

`ress = Array(res)`
`destroy!(ctx)`
### compare/verify results
`res0 = (psic.*val).*psi[perm];`
`norm( ress - res0 )`
------------------------------------------------------------------------------
------------------------------------------------------------------------------
# B U T
As I pointed out in the comments, the number of blocks has to be exactly 100. Otherwise I cannot run 
`ress = Array(res)`
to collect data from GPU, it gives me "CUDA error: an illegal memory access was encountered" error.
The entire error message reads:
------------------------------------------------------------------------------
CUDA error: an illegal memory access was encountered (code #700, ERROR_ILLEGAL_ADDRESS)
macro expansion at base.jl:148 [inlined]
#download!#5(::Bool, ::Function, ::Ptr{Complex{Float64}}, ::CUDAdrv.Mem.Buffer, ::Int64, ::CUDAdrv.CuStream) at memory.jl:224
copy!(::Array{Complex{Float64},1}, ::CUDAdrv.CuArray{Complex{Float64},1}) at array.jl:153
Array(::CUDAdrv.CuArray{Complex{Float64},1}) at array.jl:231
include_string(::String, ::String) at loading.jl:522
include_string(::String, ::String, ::Int64) at eval.jl:30
include_string(::Module, ::String, ::String, ::Int64, ::Vararg{Int64,N} where N) at eval.jl:34
(::Atom.##100#105{String,Int64,String})() at eval.jl:75
withpath(::Atom.##100#105{String,Int64,String}, ::String) at utils.jl:30
withpath(::Function, ::String) at eval.jl:38
hideprompt(::Atom.##99#104{String,Int64,String}) at repl.jl:63
macro expansion at eval.jl:73 [inlined]
(::Atom.##98#103{Dict{String,Any}})() at task.jl:80
------------------------------------------------------------------------------

# Question: what is going on here ?!

------------------------------------------------------------------------------
------------------------------------------------------------------------------

# PS: device parameters of my graphics card:
Device name : 
Quadro P4000
Total memory : 8508145664
MAX_THREADS_PER_BLOCK = 1024
MAX_BLOCK_DIM_X = 1024
MAX_BLOCK_DIM_Y = 1024
MAX_BLOCK_DIM_Z = 64
MAX_GRID_DIM_X = 2147483647
MAX_GRID_DIM_Y = 65535
MAX_GRID_DIM_Z = 65535
MAX_SHARED_MEMORY_PER_BLOCK = 49152
TOTAL_CONSTANT_MEMORY = 65536
WARP_SIZE = 32
MAX_PITCH = 2147483647
MAX_REGISTERS_PER_BLOCK = 65536
CLOCK_RATE = 1480000
TEXTURE_ALIGNMENT = 512
GPU_OVERLAP = 1
MULTIPROCESSOR_COUNT = 14
KERNEL_EXEC_TIMEOUT = 0
INTEGRATED = 0
CAN_MAP_HOST_MEMORY = 1
COMPUTE_MODE = 0
MAXIMUM_TEXTURE1D_WIDTH = 131072
MAXIMUM_TEXTURE2D_WIDTH = 131072
MAXIMUM_TEXTURE2D_HEIGHT = 65536
MAXIMUM_TEXTURE3D_WIDTH = 16384
MAXIMUM_TEXTURE3D_HEIGHT = 16384
MAXIMUM_TEXTURE3D_DEPTH = 16384
MAXIMUM_TEXTURE2D_LAYERED_WIDTH = 32768
MAXIMUM_TEXTURE2D_LAYERED_HEIGHT = 32768
MAXIMUM_TEXTURE2D_LAYERED_LAYERS = 2048
SURFACE_ALIGNMENT = 512
CONCURRENT_KERNELS = 1
ECC_ENABLED = 0
PCI_BUS_ID = 2
PCI_DEVICE_ID = 0
TCC_DRIVER = 0
MEMORY_CLOCK_RATE = 3802000
GLOBAL_MEMORY_BUS_WIDTH = 256
L2_CACHE_SIZE = 2097152
MAX_THREADS_PER_MULTIPROCESSOR = 2048
ASYNC_ENGINE_COUNT = 2
UNIFIED_ADDRESSING = 1
MAXIMUM_TEXTURE1D_LAYERED_WIDTH = 32768
MAXIMUM_TEXTURE1D_LAYERED_LAYERS = 2048
CAN_TEX2D_GATHER = 1
MAXIMUM_TEXTURE2D_GATHER_WIDTH = 32768
MAXIMUM_TEXTURE2D_GATHER_HEIGHT = 32768
MAXIMUM_TEXTURE3D_WIDTH_ALTERNATE = 8192
MAXIMUM_TEXTURE3D_HEIGHT_ALTERNATE = 8192
MAXIMUM_TEXTURE3D_DEPTH_ALTERNATE = 32768
PCI_DOMAIN_ID = 0
TEXTURE_PITCH_ALIGNMENT = 32
MAXIMUM_TEXTURECUBEMAP_WIDTH = 32768
MAXIMUM_TEXTURECUBEMAP_LAYERED_WIDTH = 32768
MAXIMUM_TEXTURECUBEMAP_LAYERED_LAYERS = 2046
MAXIMUM_SURFACE1D_WIDTH = 32768
MAXIMUM_SURFACE2D_WIDTH = 131072
MAXIMUM_SURFACE2D_HEIGHT = 65536
MAXIMUM_SURFACE3D_WIDTH = 16384
MAXIMUM_SURFACE3D_HEIGHT = 16384
MAXIMUM_SURFACE3D_DEPTH = 16384
MAXIMUM_SURFACE1D_LAYERED_WIDTH = 32768
MAXIMUM_SURFACE1D_LAYERED_LAYERS = 2048
MAXIMUM_SURFACE2D_LAYERED_WIDTH = 32768
MAXIMUM_SURFACE2D_LAYERED_HEIGHT = 32768
MAXIMUM_SURFACE2D_LAYERED_LAYERS = 2048
MAXIMUM_SURFACECUBEMAP_WIDTH = 32768
MAXIMUM_SURFACECUBEMAP_LAYERED_WIDTH = 32768
MAXIMUM_SURFACECUBEMAP_LAYERED_LAYERS = 2046
MAXIMUM_TEXTURE1D_LINEAR_WIDTH = 134217728
MAXIMUM_TEXTURE2D_LINEAR_WIDTH = 131072
MAXIMUM_TEXTURE2D_LINEAR_HEIGHT = 65000
MAXIMUM_TEXTURE2D_LINEAR_PITCH = 2097120
MAXIMUM_TEXTURE2D_MIPMAPPED_WIDTH = 32768
MAXIMUM_TEXTURE2D_MIPMAPPED_HEIGHT = 32768
COMPUTE_CAPABILITY_MAJOR = 6
COMPUTE_CAPABILITY_MINOR = 1
MAXIMUM_TEXTURE1D_MIPMAPPED_WIDTH = 16384
STREAM_PRIORITIES_SUPPORTED = 1
GLOBAL_L1_CACHE_SUPPORTED = 1
LOCAL_L1_CACHE_SUPPORTED = 1
MAX_SHARED_MEMORY_PER_MULTIPROCESSOR = 98304
MAX_REGISTERS_PER_MULTIPROCESSOR = 65536
MANAGED_MEMORY = 1
MULTI_GPU_BOARD = 0
MULTI_GPU_BOARD_GROUP_ID = 0

**Comment by algorithmx** [comment]
sorry about the ugly formatting...

**Comment by algorithmx** [comment]
Update:
if I directly rune
`@time @cuda (128,1024) kernel_zip_V1(psi_dev, psiconj_dev, val_dev, perm_dev, res)`
then followed by
`ress2 = Array(res)`
I got error:
CUDA error: an illegal memory access was encountered (code #700, ERROR_ILLEGAL_ADDRESS)
macro expansion at base.jl:148 [inlined]
#download!#5(::Bool, ::Function, ::Ptr{Complex{Float64}}, ::CUDAdrv.Mem.Buffer, ::Int64, ::CUDAdrv.CuStream) at memory.jl:224
copy!(::Array{Complex{Float64},1}, ::CUDAdrv.CuArray{Complex{Float64},1}) at array.jl:153
Array(::CUDAdrv.CuArray{Complex{Float64},1}) at array.jl:231
include_string(::String, ::String) at loading.jl:522
include_string(::String, ::String, ::Int64) at eval.jl:30
include_string(::Module, ::String, ::String, ::Int64, ::Vararg{Int64,N} where N) at eval.jl:34
(::Atom.##100#105{String,Int64,String})() at eval.jl:75
withpath(::Atom.##100#105{String,Int64,String}, ::String) at utils.jl:30
withpath(::Function, ::String) at eval.jl:38
hideprompt(::Atom.##99#104{String,Int64,String}) at repl.jl:63
macro expansion at eval.jl:73 [inlined]
(::Atom.##98#103{Dict{String,Any}})() at task.jl:80
--------------------
The function `download!` in CUDAdrv throws this error when calling lincuda api `cuMemcpyDtoH_v2`

**Comment by maleadt** [comment]
Please don't format your posts like that. Also, this isn't really an issue.
You can use the [Discourse forum](https://discourse.julialang.org/c/domain/gpu) for support questions like this 🙂 

> The function download! in CUDAdrv throws this error when calling lincuda api cuMemcpyDtoH_v2

No, calling `download!` returns the errors from previous asynchronous API calls, such as the launch of the kernel.

> the number of blocks has to be exactly 100.

Isn't that expected? You allocate arrays of `100*len` or `100*1024` items, so your kernel can only access up to that many elements. Given your index calculation (`(blockIdx().x-1) * blockDim().x + threadIdx().x`) you can only launch 100 blocks of 1024 kernels before going out of bounds, and performing an illegal memory access.

The fact that the error sticky, ie. you can't do other computations after encountering that error, is a CUDA restriction.

**Comment by algorithmx** [comment]
@maleadt Thanks, I also got a solution after lunch. I should have checked the value of i , iff it is less than the bound of the arrays then compute, else do nothing. 

In "CUDA by example" (2012) page 65, the following codes are mentioned:
```
if (tid < N)
      c[tid] = a[tid] + b[tid]
```
there the author was discussing vector add. 
I should do exactly the same thing !!

About the format, I tried to edit and this was the best result :(
